### ISSUE TITLE
Incorrect cursor position on first expansion of function/dynamic nodes

### ISSUE THREAD

**Issue by jedrzejboczar** [L3MON4D3/LuaSnip#356]
I noticed that when expanding a snippet the cursor position is on the last line of a snippet. The behaviour must have changed recently, as my snippet that relied on `luasnip.util.util.get_current_line_to_cursor` stopped working. When the function/dynamic node depends on other node, then the following expansions have correct cursor position. Do you know what might be the reason?

I tested on luasnip revision `4ecaedb`. To reproduce use the following minimal config `luasnip_cursor.lua`:
```lua
vim.cmd 'set runtimepath+=~/.local/share/nvim/site/pack/packer/start/LuaSnip'

local ls = require('luasnip')
local util = require('luasnip.util.util')
local fmt = require('luasnip.extras.fmt').fmt
local snip = ls.snippet
local txt = ls.text_node
local ins = ls.insert_node
local fun = ls.function_node

ls.config.setup {}
vim.keymap.set({'i', 'v', 'n'}, '<c-k>', ls.expand_or_jump)

ls.add_snippets('lua', {
    snip('fun', fmt([[
        function {}()
            return '{}'
        end
    ]], {
            ins(1, 'name'),
            fun(function()
                local line = util.get_current_line_to_cursor()
                local cursor = vim.api.nvim_win_get_cursor(0)
                return string.format('line = "%s", cursor = %s', line, vim.inspect(cursor))
            end, {}),
        }))
})
```

then run `nvim --noplugin -u luasnip_cursor.lua /tmp/test.lua` and type `fun` followed by `<c-k>` to expand.
The output is:
```lua
function name()
    return 'line = "end", cursor = { 3, 3 }'
end
```
If the `fun` node is changed to depend on `ins` node, then following expansions will have cursor line 1.

**Comment by L3MON4D3** [comment]
Ah, I see that's a bit unfortunate :/
I've no idea why depending on another node makes a difference (which it somehow doesn't for me, both versions show `{3,3}`)

Anyway, I don't think the cursor-position can be relied upon while the snippet is expanding (did I/the docs say something like that? :grimacing:), what's your exact usecase? I'm sure we can find a "proper" way to do this :)

**Comment by jedrzejboczar** [comment]
I have a snippet for lua that uses dynamic node to expand either to (`{}` marks insert nodes):
```
function({1})
    {0}
end
```
or 
```
function {1}({2})
    {0}
end
```
depending on what was the text on the line that we expand on. So e.g. if it's line beginning (`^%s*$`) it expands to the second variant, but for "inline" functions, like `vim.tbl_map(function...` or `local x = function...` it's the first variant.

**Comment by L3MON4D3** [comment]
Okay, one way to go about this would be [`condition`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets) (can't link exact bulletpoint :( ), imo the snippets are different enough for that to be ok, but ymmv.
Alternatively, I've been thinking about adding a `pre_expand`-event (like `enter`/`leave`), that could be used to set some value in the snippet, which can then be accessed in the `functionNode`.

**Comment by jedrzejboczar** [comment]
Ok, I think I'll try using 2 snippets with condition to avoid complicating luasnip core.

I guess I'll have to rethink my approach to snippets with the same trigger. Until now I was e.g. merging C/C++ manually and removing duplicated version ("overwriting" C with C++). Without it I was getting duplicates in nvim-cmp completion which was annoying. But I think this should be possible to solve via condition+show_condition.

**Comment by L3MON4D3** [comment]
> to avoid complicating luasnip core.

Oh don't worry about that, we'll probably end up adding this either way :D

What do you think about showing the snippet's `name` instead of its `trigger` in cmp? One snippet could be named `c-if`, the other `cpp-if`.

Another idea: extend `cpp` with `c`, but use some filetype `c_common` and extend both `c` and `cpp` with it. Then you could put snippets common to `c` and `cpp` in `c_common` and those you'd override manually in `cpp` only into `c`.

**Comment by jedrzejboczar** [comment]
> What do you think about showing the snippet's `name` instead of its `trigger` in cmp? One snippet could be named `c-if`, the other `cpp-if`.

I think trigger makes more sense, as that's what you actually typed. I would probably be confused seeing something else there. And for that specific case I'd rather not even see the C version when editing  C++ (this "overwriting" behaviour).
 
> Another idea: extend `cpp` with `c`, but use some filetype `c_common` and extend both `c` and `cpp` with it. Then you could put snippets common to `c` and `cpp` in `c_common` and those you'd override manually in `cpp` only into `c`.

Seems nice, gotta give it a try. A bit more verbose (3 separate filetypes instead of 2), but probably more straightforward. It actually sounds like yet another incarnation of "composition over inheritance" :D

**Comment by jedrzejboczar** [comment]
I tried the "common virtual filetype" approach and it works pretty well, thanks for advice.
I was also able to solve the issue with checking line before cursor, I used [something like this](https://gist.github.com/jedrzejboczar/6ceda0634e1572ac05ad4ce82a72ea3e). 

One minor thing that could (maybe) be helpful in the `show_condition` would be to have access to the snippet trigger (needed to compute the text before cursor unrelated to the snippet), currently I pass it explicitly (`snip('fu', ..., { show_condition = is_inline_function('fu') })`), but it's not a big deal.
